snippet_parser/core_test.py

```diff
@@ -146,5 +146,63 @@ def test_no_wikicode_in_section_titles(self):
             "== Section title ''with Wikicode'' == \n\nIrrelevant content{{cn}}")
         self.assertEqual(section, 'Section title with Wikicode')
 
+    def _do_extract_sentences(self, wikitext, sentences):
+        for sentence in sentences:
+            start = wikitext.find(sentence)
+            if start < 0: continue
+            marked = core._SENTENCE_MARKER_MARKUP.format(sent = sentence)
+            wikitext = wikitext[:start] + marked + wikitext[start+len(sentence):]
+        self._set_wikipedia_parse_response(wikitext)
+        ret = self._sp.extract_from_sentences(wikitext, sentences)
+        call_args = None
+        if self._wp.parse.call_count:
+            call_args = self._wp.parse.call_args[0][0]
+        ret = ret[0] if ret else []
+        return call_args, ret
+
+    def test_sentence_to_snippet(self):
+        sentences = ['A sentence.']
+        wikitext = '<p>Stuff. A sentence. More stuff.</p>'
+        _, [_, snippets]  = self._do_extract_sentences(wikitext, sentences)
+        expected = '<div class="%s"><p>Stuff. ' % (core.SNIPPET_WRAPPER_CLASS
+            ) + core._SENTENCE_MARKER_MARKUP.format(sent = 'A sentence.'
+            ) + ' More stuff.</p></div>'
+        self.assertEqual(expected, snippets[0])
+
+    def test_sentence_not_found(self):
+        sentences = ['A sentence.']
+        wikitext = "<p>Missing the sentence in the Wikitext.</p>"
+        _, [_, snippets] = self._do_extract_sentences(wikitext, sentences)
+        self.assertFalse(snippets)
+
+    def test_lengthy_sentence(self):
+        sentences = ['A '+'very '*1000+'long sentence.']
+        wikitext = '<p>Stuff. A '+'very '*1000+'long sentence. More stuff.</p>'
+        _, [_, snippets] = self._do_extract_sentences(wikitext, sentences)
+        self.assertFalse(snippets)
+
+    def test_sentence_in_list(self):
+        sentences = ['Element 3']
+        wikitext = '''
+            <p>The following is a list of elements:</p>
+            <ul>
+                <li>Element 1</li>
+                <li>Element 2</li>
+                <li>Element 3</li>
+                <li>Element 4</li>
+                <li>Element 5</li>
+            </ul>
+        '''
+        _, [_, snippets]  = self._do_extract_sentences(wikitext, sentences)
+        self.assertEqual(len(snippets), 1)
+        s = snippets[0]
+        self.assertIn('<p>The following', s)
+        self.assertNotIn('<li>Element 1', s)
+        self.assertIn('<li>Element 2', s)
+        self.assertIn(core.SENTENCE_MARKER_CLASS, s)
+        self.assertIn(core._SENTENCE_MARKER_MARKUP.format(sent = 'Element 3'), s)
+        self.assertIn('<li>Element 4', s)
+        self.assertNotIn('<li>Element 5', s)
+
```

Comment:
The tests are very nice, thank you! Just a couple of other things that stood out:

- None of the tests pass more than one sentence to to `_extract_sentences`, which of course can totally happen in the real world when there's more than one sentence per article. That's something we should probably exercise in a test too, right?
- I do think it makes sense to add a simple test for repeated sentences, to show that we intentionally return no snippets in that case, instead of (for example) crashing. Without that, it's easy to forget about this corner case when changing the code in the future and break it, even if it behaves fine right now.

Reply:
Hi @eggpi,

About testing more than one sentence to `_extract_sentences`. I passed two sentences **in two different paragraphs under the same section**, I found it only returns one snippet (one paragraph and one sentence returned). 

Then I saw the code [this line](https://github.com/AikoChou/citationhunt/blob/master/snippet_parser/core.py#L231). It seems we always get only one root in the `snippet_roots` list even if we have multiple markers? I'm thinking why it's not `snippet_roots.append(self._make_snippet_root(root))` ?

And about the test for repeated sentences, I was thinking wrong. Instead of returning no snippets, it would return a snippet with nested markers. Because the code always matches the sentence it first meets, the modified wikitext has nested markers like:
```
Stuff. <span class="sentence-marker"><span class="sentence-marker">Sentence 1</span></span> \n\nMore stuff. Sentence 1
```
So, now we just add a test to show that the return snippet has nested markers?